src/app/modules/form/admin-form/admin-form.controller.ts

```diff
@@ -36,3 +40,106 @@ export const handleListDashboardForms: RequestHandler = async (req, res) => {
+export const handleCountFormSubmissions: RequestHandler<
+  { formId: string },
+  unknown,
+  unknown,
+  { startDate?: string; endDate?: string }
+> = async (req, res) => {
+  const { formId } = req.params
+  const { startDate, endDate } = req.query
+  const sessionUserId = (req.session as Express.AuthedSession).user._id
+
+  // Step 1: Retrieve currently logged in user.
+  const adminResult = await UserService.getPopulatedUserById(sessionUserId)
+  // Step 1a: Error retrieving logged in user.
+  if (adminResult.isErr()) {
+    logger.error({
+      message: 'Error occurred whilst retrieving user',
+      meta: {
+        action: 'handleCountFormSubmissions',
+        userId: sessionUserId,
+      },
+      error: adminResult.error,
+    })
+
+    const { errorMessage, statusCode } = mapRouteError(adminResult.error)
+    return res.status(statusCode).json({ message: errorMessage })
+  }
+  // Step 1b: Successfully retrieved logged in user.
+  const admin = adminResult.value
+
+  // Step 2: Retrieve full form.
+  const formResult = await FormService.retrieveFullFormById(formId)
+  // Step 2a: Error retrieving form.
+  if (formResult.isErr()) {
+    logger.error({
+      message: 'Failed to retrieve form',
+      meta: {
+        action: 'handleCountFormSubmissions',
+        ...createReqMeta(req),
+        formId,
+      },
+      error: formResult.error,
+    })
+    const { errorMessage, statusCode } = mapRouteError(formResult.error)
+    return res.status(statusCode).json({ message: errorMessage })
+  }
+  // Step 2b: Successfully retrieved form.
+  const form = formResult.value
+
+  // Step 3: Check form permissions.
+  const permissionResult = assertHasReadPermissions(admin, form)
+  // Step 3a: Read permission error.
+  if (permissionResult.isErr()) {
+    logger.error({
+      message: 'User does not have read permissions',
+      meta: {
+        action: 'handleCountFormSubmissions',
+        ...createReqMeta(req),
+        userId: sessionUserId,
+        formId,
+      },
+      error: permissionResult.error,
+    })
+    const { errorMessage, statusCode } = mapRouteError(permissionResult.error)
+    return res.status(statusCode).json({ message: errorMessage })
+  }
+
+  // Step 4: has permissions, continue to retrieve submission counts.
+  const countResult = await SubmissionService.getFormSubmissionsCount(
+    String(form._id),
+    { startDate, endDate },
+  )
+  // Step 4a: Error retrieving form submissions counts.
+  if (countResult.isErr()) {
+    logger.error({
+      message: 'Error retrieving form submission count',
+      meta: {
+        action: 'handleCountFormSubmissions',
+        ...createReqMeta(req),
+        userId: sessionUserId,
+        formId,
+      },
+      error: countResult.error,
+    })
+    const { errorMessage, statusCode } = mapRouteError(countResult.error)
+    return res.status(statusCode).json({ message: errorMessage })
+  }
+
+  // Successfully retrieved count.
+  return res.json(countResult.value)
```

Comment:
It's different from try-catch though. To reach the next step, you have to actively unwrap the `Result` unlike in try catch where devs may forget to throw.

Doing it this way allows for better error logs, though it also makes sense to chain them and map/mapErr since the service themselves already logs errors

Thoughts, @liangyuanruo?

Reply:
yeah imo it's fine for the logs to get more generic at higher and higher layers, as long as each layer is responsible for logging the error. I think it's better to have cleaner code here by taking advantage of chaining, rather than having duplicative log messages between the controller and service.

Reply:
We just go back and forth, the initial implementation was chaining, then it became distinct steps, then etc etc HAHAHA. I'll go with this for now, can always open a PR to chain it again. Too many rebases now :P

Reply:
yup this is ok with me too, as long as the `Result` type is narrowed by testing with `isErr()`. I wouldn't want to be too opinionated here.